File: YseEngine/patcher/patcher.cpp
#include "patcher.hpp"
#include "patcherImplementation.h"
#include "pRegistry.h"

using namespace YSE;

patcher::patcher()
  : pimpl(nullptr)
{

}

patcher::~patcher() {
  if (pimpl != nullptr) {
    if (pimpl->controlledBySound) {
      pimpl->head.store(nullptr);
    }
    else {
      delete pimpl;
    }
  }
}

void patcher::create(int mainOutputs) {
  if (pimpl != nullptr) return;
  pimpl = new PATCHER::patcherImplementation(mainOutputs, this);
}

pHandle * patcher::CreateObject(const std::string & type, const std::string & args) {
  if (pimpl == nullptr) return nullptr;
  return pimpl->CreateObject(type, args);
}

void patcher::DeleteObject(pHandle * obj) {
  if (pimpl == nullptr) return;
  pimpl->DeleteObject(obj);
}

void patcher::Clear() {
  if (pimpl == nullptr) return;
  pimpl->Clear();
}

void patcher::Connect(pHandle * from, int outlet, pHandle * to, int inlet) {
  if (pimpl == nullptr) return;
  pimpl->Connect(from, outlet, to, inlet);
}

void patcher::Disconnect(pHandle * from, int outlet, pHandle * to, int inlet) {
  if (pimpl == nullptr) return;
  pimpl->Disconnect(from, outlet, to, inlet);
}

bool patcher::IsValidObject(const char * type) {
  return YSE::PATCHER::Register().IsValidObject(type);
}

std::string patcher::DumpJSON() {
  if (pimpl == nullptr) return "";
  std::string result = pimpl->DumpJSON();
  return result;
}

void patcher::ParseJSON(const std::string & content) {
  if (pimpl == nullptr) return;
  pimpl->ParseJSON(content);
}

unsigned int patcher::Objects() {
  if (pimpl == nullptr) return 0;
  return pimpl->Objects();
}

YSE::pHandle * patcher::GetHandleFromList(unsigned int obj) {
  if (pimpl == nullptr) return nullptr;
  return pimpl->GetHandleFromList(obj);
}

YSE::pHandle * patcher::GetHandleFromID(unsigned int objID) {
  if (pimpl == nullptr) return nullptr;
  return pimpl->GetHandleFromID(objID);
}

void patcher::PassBang(const std::string & to) {
  if (pimpl == nullptr) return;
  pimpl->PassBang(to, T_GUI);
}

void patcher::PassData(int value, const std::string & to) {
  if (pimpl == nullptr) return;
  pimpl->PassData(value, to, T_GUI);
}

void patcher::PassData(float value, const std::string & to) {
  if (pimpl == nullptr) return;
  pimpl->PassData(value, to, T_GUI);
}

void patcher::PassData(const std::string & value, const std::string & to) {
  if (pimpl == nullptr) return;
  pimpl->PassData(value, to, T_GUI);
}

File: PCLYSE/Patcher.cs
using System;
using System.Collections.Generic;
using System.Text;
using IYse;
using YSE;

namespace YSE
{
	class OscHandler : Yse.oscHandler
	{
		public override void Send(string to)
		{
			OnOscBang(to);
		}

		public override void Send(string to, float value)
		{
			OnOscFloat(to, value);
		}

		public override void Send(string to, int value)
		{
			OnOscInt(to, value);
		}

		public override void Send(string to, string value)
		{
			OnOscString(to, value);
		}

		public event IYse.OnOscBangHandler OnOscBang;
		public event IYse.OnOscIntHandler OnOscInt;
		public event IYse.OnOscFloatHandler OnOscFloat;
		public event IYse.OnOscStringHandler OnOscString;
	}

  public class Patcher : Yse.patcher, IYse.IPatcher
  {
		public event OnOscBangHandler OnOscBang;
		public event OnOscFloatHandler OnOscFloat;
		public event OnOscIntHandler OnOscInt;
		public event OnOscStringHandler OnOscString;

		private OscHandler Osc;

		public Patcher()
		{
			Osc = new OscHandler();
			Osc.OnOscBang += (to) => OnOscBang(to);
			Osc.OnOscInt += (to, value) => OnOscInt(to, value);
			Osc.OnOscFloat += (to, value) => OnOscFloat(to, value);
			Osc.OnOscString += (to, value) => OnOscString(to, value);
		}

		public void Connect(IYse.IHandle from, int pinOut, IYse.IHandle to, int pinIn)
    {
      base.Connect(((Handle)from).GetSource(), pinOut, ((Handle)to).GetSource(), pinIn);
    }

    public void Create(int mainOutputs)
    {
      create(mainOutputs);
			SetOscHandler(Osc);
		}

    public void Disconnect(IYse.IHandle from, int outlet, IYse.IHandle to, int inlet)
    {
      base.Disconnect(((Handle)from).GetSource(), outlet, ((Handle)to).GetSource(), inlet);
    }

    public void DeleteObject(IYse.IHandle handle)
    {
      base.DeleteObject(((Handle)handle).GetSource());
    }

    IYse.IHandle IYse.IPatcher.CreateObject(string type, string args)
    {
      Yse.pHandle handle = base.CreateObject(type, args);
      if (handle == null) return null;
      else return new Handle(handle);
    }

    public uint NumObjects()
    {
      return base.Objects();
    }

    IYse.IHandle IYse.IPatcher.GetHandleFromList(uint obj)
    {
      Yse.pHandle handle = base.GetHandleFromList(obj);
      if (handle == null) return null;
      else return new Handle(handle);
    }

    IYse.IHandle IYse.IPatcher.GetHandleFromID(uint obj)
    {
      Yse.pHandle handle = base.GetHandleFromID(obj);
      if (handle == null) return null;
      else return new Handle(handle);
    }



  }
}
